pycbc/vetoes/chisq_cpu.pyx

```diff
@@ -0,0 +1,145 @@
+                     int blen):
+    # Do I need to declare UINT vs INT??
+    cdef int num_parallel_regions, bstart, bend, i, j, k, r, start, end
+    cdef numpy.ndarray[REALTYPE, ndim=1] outr, outi, pr, pi, vsr, vsi, outr_tmp, outi_tmp
+    cdef COMPLEXTYPE v
+    cdef REALTYPE vr, vi, t1, t2, k1, k2, k3, vs, va
+
+    #real_type = real_same_precision_as(chisq)
+    real_type = numpy.float64
+    #complex_type = complex_same_precision_as(chisq)
+
+    num_parallel_regions = 16
+
+    outr = numpy.zeros(n, dtype=real_type)
+    outi = numpy.zeros(n, dtype=real_type)
+    pr = numpy.zeros(n, dtype=real_type)
+    pi = numpy.zeros(n, dtype=real_type)
+    vsr = numpy.zeros(n, dtype=real_type)
+    vsi = numpy.zeros(n, dtype=real_type)
+    outr_tmp = numpy.zeros(n, dtype=real_type)
+    outi_tmp = numpy.zeros(n, dtype=real_type)
+
+    for r in range(blen):
+        bstart = bins[r] # int
+        bend = bins[r+1] # int
+        blen = bend - bstart # int
+
+        #outr = numpy.zeros(n, dtype=real_type)
+        #outi = numpy.zeros(n, dtype=real_type)
+        for i in range(n):
+            outr[i] = 0.
+            outi[i] = 0.
+
+
+        # CAN WE MAKE THIS LOOP PARALLEL?!
+        for k in range(num_parallel_regions):
+            start = blen * k / num_parallel_regions + bstart # uint
+            end = blen * (k + 1) / num_parallel_regions + bstart # uint
+
+            # start the cumulative rotations at the offset point
+            #pr = numpy.zeros(n, dtype=real_type)
+            #pi = numpy.zeros(n, dtype=real_type)
+            #vsr = numpy.zeros(n, dtype=real_type)
+            #vsi = numpy.zeros(n, dtype=real_type)
+            #outr_tmp = numpy.zeros(n, dtype=real_type)
+            #outi_tmp = numpy.zeros(n, dtype=real_type)
+            
+
+            for i in range(n):
+                pr[i] = cos(2 * 3.141592653 * shifts[i] * (start) / slen)
+                pi[i] = sin(2 * 3.141592653 * shifts[i] * (start) / slen)
+                vsr[i] = cos(2 * 3.141592653 * shifts[i] / slen)
+                vsi[i] = sin(2 * 3.141592653 * shifts[i] / slen)
+                outr_tmp[i] = 0
+                outi_tmp[i] = 0
+
+            for j in range(start, end): # uint
+                v = v1[j]
+                vr = v.real
+                vi = v.imag
+                vs = vr + vi;
+                va = vi - vr;
+
+                for i in range(n):
+                    t1 = pr[i]
+                    t2 = pi[i]
+
+                    # Complex multiply pr[i] * v
+                    k1 = vr * (t1 + t2)
+                    k2 = t1 * va
+                    k3 = t2 * vs
+
+                    outr_tmp[i] += k1 - k3
+                    outi_tmp[i] += k1 + k2
+
+                    # phase shift for the next time point
+                    pr[i] = t1 * vsr[i] - t2 * vsi[i]
+                    pi[i] = t1 * vsi[i] + t2 * vsr[i]
+
+            # WHAT DOES THIS MEAN?? : pragma omp critical
+            for i in range(n):
```

Comment:
In a parallel code this ensures that only one thread updates the variables outr[i] and out[i] at a time, otherwise you can have the case where two try to update simultaneously and you actually miss a bins result being added to the total. It can't be done this way with cython anyway, so can be simplified.

Reply:
Okay.